Win shell wip

yeff-bridges opened the pull request:
Let's see how this works! I still plan to add an object to handle command arguments, some new command related exceptions to throw, and to reduce a few dependencies between the new classes if possible. I also am not sure exactly how the hyperlinks worked, so I don't know what I did to disable them. It's likely a small fix, so let's see what this pull request will let you do!

yeff-bridges committed:
Primitive Shell uploaded

There are still many features to add and changes to make!

yeff-bridges committed:
Delete Shell.c

yeff-bridges committed:
Shell 1.2 added

yeff-bridges committed:
Command Parser and Executor WIP DO NOT PULL

yeff-bridges committed:
CommandParser, CommandExecutor, SingleProcessCommand, MultiprocessCommand, and BuiltinLibrary skeleton classes all added. Code runs exactly like before because none of these classes are yet used.

yeff-bridges committed:
Working protoype of the new Command Processor, including CommandExecutor, CommandParser, and Multi & Single Process Command objects. No new functionality has been added, but this new setup is more clean and extendable. Any ideas to clean it futher and reduce dependancies are welcome.

yeff-bridges committed:
New Branch added to store new command parser

DarkChocoholicDev committed:
Fix loss of hyperlink functionality.

DarkChocoholicDev committed:
Illustrate using a property instead of an old-school "getter" method.

DarkChocoholicDev committed:
Finish replacing "getter" methods with properties.

DarkChocoholicDev reviewed (approved):
I didn't figure out that this was a code review (or how to review it) until well after I had piled on to your "pull request" branch. Ooops. :D So we'll just take all the changes and continue on into the sunset.  Party on, Wayne!

DarkChocoholicDev committed:
Phase 1 (prep-work) of support for multiple output views.

1. Created UIManager to keep track of "shell sessions".
2. MainWindow is no longer passed around, except to UIManager.
3. ConsoleWindow instances provide command output functionality.
4. A ShellSession pairs a ConsoleWindow instance with a current directory.
5. The command processing layer is passed a UIManager instance so that it can manage ShellSession instances via command line commands.
6. ProcessorCommand instances now contain a ConsoleWindow instance so that the output of an individual command can be targeted at a specific output window.
7. The notion of a "default" shell session exists so that the UI can simply request command execution on the default shell session.
8. Command links specify the shell session on which they should run. Usually it is the same shell session that wrote the command link to an output window, but it doesn't have to be that way.

For this check-in, only a single shell session is created. If all goes well, there should be no perceptible functionality difference. :D But just wait until Phase 2!

yeff-bridges committed:
Fixed minor bug where window doesn't update before outputing a message. Setup work finished to start implementing commands that work with multiple threads. This commit is not a working model and is being pushed so that I may pull it on my laptop later tonight.

yeff-bridges committed:
Working version that runs single commands with one set of arguments. Ready for addition of DLL's.

DarkChocoholicDev committed:
Initial updates to support I/O redirection for external processes.

1. Add "cls" (clear screen) command to clear the console output.

2. Use extended version of Process.Start() to redirect the standard input/output/error streams for launched processes to that they stream to/from the WinShell console by default. Blocking I/O runs on separate threads.

3. Adjust console I/O routines so that they run on the UI thread when necessary. That way, the caller doesn't need to care about which thread is running.

4. Initial/prototype draft code for translating "key" presses into characters to be redirected to external processes. This code was pulled from a StackOverflow sample and updated as needed, though it still needs further refinement and documentation.

yeff-bridges committed:
Overhauled the command processing system to support external command libraries. In doing so, the following changes have been made:

1. The old system of using "ProcessorCommand" objects has been completely scrapped. This was replaced with a system of "CommandLibrary" and "CommandHandler" objects. The "CommandLibrary" stores information about each command, such as its name, description and the Handler object that executes it, and this information is used by the CommandParser, CommandHandler, and LibraryManager objects to execute the commands upon request. The "CommandHandler" object includes a required ExecuteCommand method that handles each command passed to it based on the CommandDescriptor it is passed.

2. A LibraryManger object has been added as a main component to the CommandProcessor to server two distinct puroposes. The first is to initialize and load all libraries of commands provided by the user (more on this later). The second is to provide an interface with which the rest of the shell may interact with the command libraries, and vise versa.

3. The old "BuiltinLibrary" class has been scrapped, with its duties broken up between the CommandExecutor, BuiltinHandler, and new BuiltinLibrary objects.

4. An App.config file is now being used by the LibraryManager to locate any external libraries that the user wishes to include, if any at all. The "MyDlls" is used to store the file path to a directory containing various compiled command dll's to be used by the shell, with a default value of "null" indicating that there are no external libraries to be loaded. I've included my config file as a default for now.

Note that while basic error handling has been implemented, much is left to be desired. The Shell won't likely crash in the face of a user error, but it won't always be that helpful either.
